File: tryout/backend/api/src/main/scala/kpn/api/custom/ScopedRouteType.scala
package kpn.api.custom

import kpn.api.common.RouteScope
import kpn.api.common.RouteType

object ScopedRouteType {

  val rwn: ScopedRouteType = ScopedRouteType(RouteType.hiking, RouteScope.regional)
  val rcn: ScopedRouteType = ScopedRouteType(RouteType.cycling, RouteScope.regional)
  val rmn: ScopedRouteType = ScopedRouteType(RouteType.motorboat, RouteScope.regional)
  val lwn: ScopedRouteType = ScopedRouteType(RouteType.hiking, RouteScope.local)
  val lcn: ScopedRouteType = ScopedRouteType(RouteType.cycling, RouteScope.local)
  val lpn: ScopedRouteType = ScopedRouteType(RouteType.canoe, RouteScope.local)

  def apply(routeType: RouteType, routeScope: RouteScope): ScopedRouteType = {
    val routeTypeLetter = RouteTypeLetter.letter(routeType)
    val routeScopeLetter = RouteScopeLetter.letter(routeScope)
    val key = s"$routeScopeLetter${routeTypeLetter}n"
    ScopedRouteType(routeType, routeScope, key)
  }

  val all: Seq[ScopedRouteType] = {
    RouteType.values.flatMap { routeType =>
      RouteScope.all.map(scope => ScopedRouteType(routeType, scope))
    }
  }

  def withKey(key: String): Option[ScopedRouteType] = {
    all.find(_.key == key)
  }

  def from(routeType: RouteType, routeScope: RouteScope): ScopedRouteType = {
    all.find(ns => ns.routeType == routeType && ns.routeScope == routeScope).get
  }
}

case class ScopedRouteType(routeType: RouteType, routeScope: RouteScope, key: String) {

  override def toString: String = key

  def nodeRefTagKey: String = s"${key}_ref"

  def nodeNameTagKey: String = s"${key}_name"

  def proposedNodeRefTagKey: String = s"proposed:${key}_ref"

  def proposedNodeNameTagKey: String = s"proposed:${key}_name"

  def expectedRouteRelationsTag: String = s"expected_${key}_route_relations"
}

File: backend/api/src/main/java/kpn/api/custom/ScopedRouteType.java
package kpn.api.custom;

import kpn.api.common.RouteScope;
import kpn.api.common.RouteType;

import java.util.stream.Stream;
import com.google.common.collect.ImmutableList;

public record ScopedRouteType(
  RouteType routeType,
  RouteScope routeScope,
  String key
) {

  public ScopedRouteType(RouteType routeType, RouteScope routeScope) {
    String routeTypeLetter = RouteTypeLetter.letter(routeType);
    String routeScopeLetter = RouteScopeLetter.letter(routeScope);
    String key = routeScopeLetter + routeTypeLetter + "n";
    this(routeType, routeScope, key);
  }

  public String nodeRefTagKey() {
    return key + "_ref";
  }

  public String nodeNameTagKey() {
    return key + "_name";
  }

  public String proposedNodeRefTagKey() {
    return "proposed:" + key + "_ref";
  }

  public String proposedNodeNameTagKey() {
    return "proposed:" + key + "_name";
  }

  public String expectedRouteRelationsTag() {
    return "expected_" + key + "_route_relations";
  }

  public static ScopedRouteType rwn = new ScopedRouteType(RouteType.HIKING, RouteScope.REGIONAL);
  public static ScopedRouteType rcn = new ScopedRouteType(RouteType.CYCLING, RouteScope.REGIONAL);
  public static ScopedRouteType rmn = new ScopedRouteType(RouteType.CANOE, RouteScope.REGIONAL);
  public static ScopedRouteType lwn = new ScopedRouteType(RouteType.HIKING, RouteScope.LOCAL);
  public static ScopedRouteType lcn = new ScopedRouteType(RouteType.CYCLING, RouteScope.LOCAL);
  public static ScopedRouteType lpn = new ScopedRouteType(RouteType.CANOE, RouteScope.LOCAL);

  public static ImmutableList<ScopedRouteType> all() {
    return Stream.of(RouteType.values()).flatMap(routeType ->
      RouteScope.all().stream().map(scope -> new ScopedRouteType(routeType, scope))
    ).collect(ImmutableList.toImmutableList());
  }

//  public static Optional<ScopedRouteType> withKey(String key) {
//    all.find(_.key == key)
//  }

//  public statuc ScopedRouteType from(RouteType routeType, RouteScope routeScope) {
//    all.find(ns => ns.routeType == routeType && ns.routeScope == routeScope).get
//  }

}
